src/slice.c

```diff
@@ -76,51 +77,82 @@ static SEXP list_slice(SEXP x, SEXP index) {
 
 #undef SLICE_BARRIER
 
+static SEXP df_slice(SEXP x, SEXP index) {
+  R_len_t n = Rf_length(x);
+  SEXP out = PROTECT(Rf_allocVector(VECSXP, n));
+
+  SEXP nms = Rf_getAttrib(x, R_NamesSymbol);
+  Rf_setAttrib(out, R_NamesSymbol, nms);
+
+  for (R_len_t i = 0; i < n; ++i) {
+    SEXP sliced = vec_slice_impl(VECTOR_ELT(x, i), index, true);
+    SET_VECTOR_ELT(out, i, sliced);
+  }
+
+  UNPROTECT(1);
+  return out;
+}
+
+
 static SEXP vec_slice_dispatch(SEXP x, SEXP index) {
-  return vctrs_dispatch2(syms_vec_slice_dispatch, fns_vec_slice_dispatch,
-                         syms_x, x,
-                         syms_i, index);
+  SEXP out = PROTECT(vctrs_dispatch2(syms_vec_slice_dispatch, fns_vec_slice_dispatch,
+                                     syms_x, x,
+                                     syms_i, index));
+  out = vctrs_restore(out, x, index);
+
+  UNPROTECT(1);
+  return out;
 }
 
-static SEXP vec_slice_switch(SEXP x, SEXP index, bool dispatch) {
+static SEXP vec_slice_impl(SEXP x, SEXP index, bool dispatch) {
+  if (has_dim(x)) {
+    return vec_slice_dispatch(x, index);
+  }
+
+  SEXP out = R_NilValue;
+
   switch (vec_typeof_impl(x, dispatch)) {
   case vctrs_type_null:
     Rf_error("Internal error: Unexpected `NULL` in `vec_slice_impl()`.");
 
   case vctrs_type_logical:
-    return lgl_slice(x, index);
+    out = lgl_slice(x, index);
+    break;
   case vctrs_type_integer:
-    return int_slice(x, index);
+    out = int_slice(x, index);
+    break;
   case vctrs_type_double:
-    return dbl_slice(x, index);
+    out = dbl_slice(x, index);
+    break;
   case vctrs_type_complex:
-    return cpl_slice(x, index);
+    out = cpl_slice(x, index);
+    break;
   case vctrs_type_character:
-    return chr_slice(x, index);
+    out = chr_slice(x, index);
+    break;
   case vctrs_type_raw:
-    return raw_slice(x, index);
+    out = raw_slice(x, index);
+    break;
   case vctrs_type_list:
-    return list_slice(x, index);
-
-  default:
-    return R_NilValue;
-  }
-}
+    out = list_slice(x, index);
+    break;
 
-static SEXP vec_slice_impl(SEXP x, SEXP index, bool dispatch) {
-  SEXP out = R_NilValue;
+  case vctrs_type_dataframe:
```

Comment:
Slice records too?

Reply:
They are not in the C types yet, I'll add them later on. I also like keeping a type fully implemented in R during the refactorings.